### PR TITLE
More handle cleanup

### LINKED ISSUE #3548: [Tracking] cleanup handle.py
- [x] ConstantObject does not reveal with the underlying value type is (#2720)
- [x] NonConstantObject type has no purpose (#2720)
- [x] ModifiableObject is both the class for logic arrays and every other value type causing unnecessary functionality to be inherited (#2720)
- [x] NonHierarchyArrayObject is both the class for non-hierarchy arrays and a base class for ModifiableObject and all more specific typed objects causing unnecessary functionality to be inherited (#2720)
- [x] Unclear which types are abstract and which aren't (#2720)
- [x] string casts on handles is the path unless it's a value handle, then it's the string cast of the value (#2720)
- [x] equality is handle equality unless it's a value handle, then it's the equality of the value of the handle and this breaks the hashability, whoops (#2720)
- [x] unnecessary specification of `__ne__` (#2720)
- [x] unnecessary access of handle attributes on handle creation (#2720)
- [ ] unclear what is public method and what isn't
- [x] `__getattr__` syntax conflicts with attribute/method access (introduces `__getitem__` #3655)
- [x] `__iter__` on HierarchyObject and HierarchyArrayObject return child handles instead of keys, like you might expect from a pseudo-dictionary/list type (introduces `_keys`, `_values`, and `_items` #3655)
- [x] `__len__` isn't defined on HierarchyObject (#3655)
- [x] `_id` takes an "extended" argument instead of just having the user add backslashes themselves like they might expect, and extended=True by default (deprecated and replaced with `__getitem__` #3655)
- [x] `_id` is a terrible name (deprecated and replaced with `__getitem__` #3655)
- [x] `_range` isn't available on HierarchyArrayObject (#3655)
- [ ] `_range` is a crappy interface, prefer `left` and `right`
- [ ] `.value` property cannot be properly typed since type systems assume getter and setter of property get and set using the same type
- [x] clean up dead code (#2720)
- [x] clean up unnecessarily inherited attributes/methods (#2720)
- [x] tons of unnecessary logs (#2720)
- [ ] fix/improve documentation

### LINKED ISSUE #3604: Integrate `Range` and `Array` types into the handle classes
Handle classes like `HierarchyArrayObject` and `NonHierarchyIndexableObjectBase` which expose information about the array's ranges could introduce properties like `range`, `left`, and `right` which leverage `cocotb.types.Range`. And `NonHierarchyIndexableObjectBase.value` could return `Array` objects to give array objects with consistent indexing back.

### LINKED ISSUE #3678: Change `ModifiableObject.drivers` and `ModifiableObject.loads` to be Python iterables
Currently they just return `cocotb.simulator.gpi_hdl_iterator` which is not a standard Python iterator, but has it's own interface and returns raw handles instead of `SimHandleBase`s.

### LINKED ISSUE #3728: Add a `simulator.is_indexable` call
Right now, `simulator.get_range` is designed to return `None` if the type isn't indexable. This means that you can't get the range of an object that isn't indexable, but does have a range (like strings and logic arrays in some simulators).

### PR DESCRIPTION
**Pull request by ktbarrett**
This should be the final PR. Closes #3548.

This PR:
* renames the value handles
* removes what was `ModifiableObject` and the non-functional `drivers` and `loads` methods (closes #3678)
* adds the `set` method to set values in a type-safe way
* improves typing and documentation

### CODE REVIEW

**Comment by ktbarrett** [comment]
So currently the issue is with `LogicObject` and how it works with multi-dimensional packed arrays and logic scalars.

# Issues with scalars

When a `logic`, `reg`, etc. in Verilog is seen, or a `std_logic` in VHDL, a `LogicObject` is used. But such an object is not indexable, doesn't have a range, and isn't iterable. However, `LogicObject` provides all that functionality, so it must be designed to work in those situations.

Currently there is an `_is_indexable` property we are checking, which should be False for scalars, which causes indexing to raise, iteration to return nothing, and range to be 1-length and whatever the defaults are (currently it's [-1, -1] from the GPI base classes).

# Issues with multi-dimensional packed arrays

When a multi-dimensional packed logic array is seen in Verilog, we can't assume we can index it, as simulators can assume that it's an indivisible object. This means that it also isn't indexable. But `LogicObject` lets you do both. In these cases `LogicObject` assumes indexing will fail, and so there is some logic in `__iter__` to convert that failure into empty iteration.

Another issue is that the length of the object is the number of total bits in the object, so a `logic[2:0][2:0]` has a length of 9. This makes sense because when getting/setting the value, you do so with the whole object at a time. But when getting the range, we only get the outer-most dimension of the range: `[2:0]`. So `len(handle) != len(handle.range)`. This means we can't set the range when getting the value appropriately.

# Solutions

IMO, we should separate scalars, single-dimension packed arrays (and `std_logic_vector` in VHDL), and multi-dimension packed arrays into separate types. ***None*** of them should be iterable or indexable.

Scalars would not have a range and getting/setting the value would use `Logic` instead of a `LogicArray`.

Single-dimension packed arrays would have a range and getting the value would return a `LogicArray` with the appropriate range set. Setting the value would take any SizedIterable of the same length (as if it were converted into `LogicArray`) or an `int`.

Multi-dimensional packed arrays would ideally return some composite range object. Getting the value would return an `Array[LogicArray]` or a custom `MultiDimensionalLogicArray` that could handle multiple-levels of indexing and also maybe setting the whole value at once. Setting would take either a flattened SizedIterable of the same length as the full object, SizedIterables of SizedIterables with the correct dimension and size (as if it were converted into `MultiDimensionalLogicArray`), or an `int`. This will require changes the GPI to return the ranges of packed objects, which seems unlikely to be supported.

Failing the above answer, multi-dimensional packed arrays could return a range of 0 to length-1, which I feel would be valid because flattened values don't necessarily have a valid range, so it can be whatever. Getting and setting the value would work like single-dimension packed arrays at that point.

The single-dimension and simple multi-dimension solution could be combined if we can ensure that at the GPI level, the range of packed objects is listed as 0 to length-1 rather than the range of the first dimension.

**Comment by ktbarrett** [comment]
I decided to split this PR up. The previous comment no longer applies to this PR.

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:324]:
```suggestion
    r"""A simulation object that is a name-indexed collection of hierarchical simulation objects.
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:347]:
```suggestion
    Accessing a non-existent child with attribute syntax results in an :class:`AttributeError`,
    and accessing a non-existent child with index syntax results in a :class:`KeyError`.
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:451]:
```suggestion
    Accessing a non-existent child results in an :class:`IndexError`.
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:615]:
```suggestion
        """Get or set the value of the simulation object.
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:699]:
```suggestion
        """Schedule a write of the given value to a simulator object.
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:728]:
```suggestion
    A particular child can be retrieved using its index:
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:1176]:
```suggestion
    This type is used when a ``string`` (VHDL or Verilog) simulation object is seen.
```

File: tests/test_cases/test_compare/test_compare.py
Comment (cmarqu) [inline tests/test_cases/test_compare/test_compare.py:58]:
```suggestion
    """Test for ValueObjectBase comparisons."""
```

**Review by cmarqu — COMMENTED** [review]
A first look, will go over it again later.

**Comment by ktbarrett** [comment]
@cmarqu Take your time. This module needs *lots* of attention wrt documentation.

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:554]:
```suggestion
    """Action used for placing a value into a given handle. This is the default action.
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:804]:
```suggestion
    This object is used whenever an array is seen that isn't either a logic array or string.
```

File: src/cocotb/handle.py
Comment (ktbarrett) [reply src/cocotb/handle.py:706]: This action can only be deposit, freeze, or release, like `_GpiSetAction`.

File: src/cocotb/handle.py
Comment (ktbarrett) [reply src/cocotb/handle.py:578]: Please recheck what I put there. I made it more explicit.

File: src/cocotb/handle.py
Comment (ktbarrett) [reply src/cocotb/handle.py:570]: Please recheck.

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:554]:
```suggestion
    """Action used for placing a value into a given handle. This is the default action.
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:572]:
```suggestion
    :class:`Freeze`\ s will act as a no-op.
```

File: src/cocotb/handle.py
Comment (cmarqu) [inline src/cocotb/handle.py:1224]: Maybe link to https://docs.python.org/3/library/codecs.html#standard-encodings ?

**Comment by cmarqu** [comment]
The inheritance diagram at https://cocotb--3733.org.readthedocs.build/en/3733/library_reference.html#simulation-object-handles is now outdated.

It could be regenerated by locally reverting the change in https://github.com/cocotb/cocotb/pull/3457/files and overwriting the SVG with the image gotten from a browser.
Or maybe Sphinx has been fixed in the meantime?

**Comment by ktbarrett** [comment]
@cmarqu It's outputting SVG, but not displaying the image, just the code. Any ideas?

**Comment by cmarqu** [comment]
> @cmarqu It's outputting SVG, but not displaying the image, just the code. Any ideas?

Works for me. Do you have a file like `cocotb/.nox/.cache/docs_out/_images/inheritance-788f556e7491e4b60b5d6114abe9f8ac3c8d40f5.svg`?
If not, is it really outputting SVG, not graphviz/dot syntax (if so, you are probably missing the `dot` tool)?

Anyway, the output looks a bit "wilder" than  before, maybe we should draw it by hand anyway...

**Comment by ktbarrett** [comment]
Yeah, it definitely needs to be done by hand. There is ABC and Generic and a few other base classes we really don't need to show. Ideally though, we wouldn't document the inheritance hierarchy and instead document only each concrete class. I don't expect many users understand inheritance at any level.

**Comment by ktbarrett** [comment]
I'm going to leave the diagram unfinished for now, as there will be another follow on commit that changes the hierarchy (bye bye IndexableValueObjectBase).

**Comment by cmarqu** [comment]
> Ideally though, we wouldn't document the inheritance hierarchy and instead document only each concrete class. I don't expect many users understand inheritance at any level.

I'm fine with just removing the diagram.

**Comment by ktbarrett** [comment]
I think as long as we are documenting inheritance, it's valuable. But if we don't, obviously it isn't.

**Comment by ktbarrett** [comment]
Considering I'm putting off the diagram update until the next PR, can I get a final review @cocotb/maintainers?

File: src/cocotb/handle.py
Comment (cmarqu) [reply src/cocotb/handle.py:1285]: Maybe because `SimHandleBase` is both a return type annotation and a reference in the "Returns:" part of the docstring?
It's not duplicated in the old version, https://docs.cocotb.org/en/latest/library_reference.html#cocotb.handle.SimHandle

Comment (ktbarrett) [reply src/cocotb/handle.py:1285]: I ran into this in other cases as well, like on the `value` getters and setters. It's a bug in Sphinx that has something to do with the cache AFAICT. I rebuild fixed the errors that I saw locally.

**Review by marlonjames — APPROVED** [review]

File: src/cocotb/handle.py
Comment (marlonjames) [reply src/cocotb/handle.py:1285]: Looks like still an issue but I'm fine with merging and seeing if it is fixed on master.

Comment (ktbarrett) [reply src/cocotb/handle.py:1285]: Yeah there's something up with the cache. I fixed the missing `"` character from your last comment, but it's still missing from the documentation. Maybe master will fix itself. I wonder if there's a way to clear or not use the cache that we can set in `.readthedocs.yml`

**Pull request merged**
